### ISSUE TITLE
int32 overflow in Triton kernels silently corrupts output (and sometimes crashes) for non-contiguous inputs at long sequence lengths

### ISSUE THREAD

**Issue by juju812** [thu-ml/SageAttention#386]
> ### Summary
>
> The Triton kernels compute element offsets in int32. With non-contiguous q/k/v — e.g.
> views into a fused QKV buffer, a very common pattern — the sequence-dim stride is
> `3 * num_heads * head_dim` rather than `head_dim`, and `offs_n * stride_in` overflows
> `2**31` at sequence lengths that are ordinary for video models.
>
> The failure mode is **silent corruption first, illegal memory access second**: once the
> offset wraps, whether it faults depends on whether the wrapped address happens to be
> mapped. Below is a self-checking repro that never faults and returns a wrong answer.
>
> ### Root cause
>
> `sageattention/triton/quant_per_thread.py`:
>
> ```python
> offs_n = off_blk * BLK + tl.arange(0, BLK // 8) * 8 + off_tld      # int32
> input_ptrs = Input + off_b * stride_iz + off_h * stride_ih \
>            + offs_n[:, None] * stride_in + offs_k[None, :]
> ```
>
> `tl.program_id` and `tl.arange` are int32, so `index * stride` is evaluated in int32
> and wraps before it is added to the (64-bit) pointer. The overflow condition is
>
> ```
> (seq - 1) * stride_in + (num_heads - 1) * stride_ih + head_dim - 1 >= 2**31
> ```
>
> Notably, the attention kernels already guard the batch and head indices:
>
> ```python
> off_z = tl.program_id(2).to(tl.int64)
> off_h = tl.program_id(1).to(tl.int64)
> ```
>
> The sequence index — the only one that actually grows — was left as int32, in those
> kernels and in every quant kernel.
>
> ### Why this has not been reported before
>
> With contiguous q/k/v the dominant term is `num_heads * seq * head_dim`, so the
> threshold sits around seq ≈ 300k and is rarely reached. Non-contiguous views multiply
> the sequence-dim stride by ~3x and bring the threshold down to ~10^5 — routine for
> long-video diffusion transformers.
>
> ### Reproduction
>
> RTX 4090 (sm89), sageattention 2.2.0, PyTorch 2.9.1+cu130, Triton 3.5.1.
> 56 heads, head_dim 128, fp16, HND, q/k/v as views into one fused buffer
> (`stride(2) == 21504`).
>
> **1. It crashes.** One sequence length per process — an illegal access poisons the
> CUDA context, and the traceback points at whichever kernel launches next, not at the
> one that faulted.
>
> ```python
> import torch
> from sageattention import sageattn
>
> seq, heads, dim = 99866, 56, 128     # 99865 passes, 99866 faults
> inner = heads * dim
> buf = torch.randn(seq, inner * 3, dtype=torch.float16, device="cuda")
> q, k, v = (t.view(seq, heads, dim).transpose(0, 1).unsqueeze(0)
>            for t in buf.split(inner, dim=-1))
> out = sageattn(q, k, v, is_causal=False, tensor_layout="HND")
> torch.cuda.synchronize()
> ```
>
> | seq | max element offset | vs `2**31` | result |
> |---:|---:|---:|---|
> | 98,304 | 2,113,914,879 | −33,568,769 | ok |
> | 99,865 | 2,147,482,623 | −1,025 | ok |
> | 99,866 | 2,147,504,127 | +20,479 | illegal memory access |
>
> The same sweep with **contiguous** tensors passes at every length we tried (up to
> 131,072) — the non-contiguous layout is required to reproduce.
>
> **2. It silently returns the wrong answer.** Put q/k/v in the upper half of a buffer
> twice the size they need: a wrapped offset (~`2**31` elements below the base) then
> lands back inside the same allocation, so it is always mapped and never faults. The
> oracle is the same call on contiguous copies of the same values — contiguity only
> shrinks the strides, it cannot change the result.
>
> ```python
> import torch
> from sageattention import sageattn_qk_int8_pv_fp8_cuda as attn
>
> seq, heads, dim = 99866, 56, 128
> inner = heads * dim
> KW = dict(tensor_layout="HND", is_causal=False, qk_quant_gran="per_thread",
>           pv_accum_dtype="fp32+fp16", smooth_k=False, smooth_v=False)
>
> torch.manual_seed(0)
> buf = torch.randn(seq * 2, inner * 3, dtype=torch.float16, device="cuda")
> q, k, v = (t.view(seq, heads, dim).transpose(0, 1).unsqueeze(0)
>            for t in buf[seq:seq * 2].split(inner, dim=-1))
>
> qc, kc, vc = (t.contiguous() for t in (q, k, v))
> ref = attn(qc, kc, vc, **KW).clone()
> del qc, kc, vc; torch.cuda.empty_cache()
>
> out = attn(q, k, v, **KW)
> torch.cuda.synchronize()
> print(torch.equal(out, ref), int((out != ref).any(-1).sum()))
> ```
>
> | seq | rows differing from the contiguous reference | error raised |
> |---:|---|---|
> | 98,304 | 0 of 5,505,024 (bit-identical) | — |
> | 99,866 | **5,471,027 of 5,592,496**, max abs diff 2.6e-2 | **none** |
>
> A single out-of-range key row is enough: the softmax denominator spreads it over 97.8%
> of the output.
>
> The Triton attention path is affected the same way — with
> `sageattn_qk_int8_pv_fp16_triton` and the same inputs, seq 99,866 gives 5,587,143
> differing rows non-causal and 205 causal.
>
> (This repro calls the sm89 entry point directly rather than `sageattn`, for two
> reasons: `sageattn()` accepts `**kwargs` and never forwards them, so `smooth_k=False`
> is silently ignored — arguably a bug of its own — and with `smooth_k=True` the oracle
> is polluted, because `k.mean()` over a strided tensor differs from the contiguous one
> by an ulp.)
>
> ### Affected files
>
> All seven files under `sageattention/triton/` compute `index * stride` in int32. In
> the varlen kernels the exposure looks worse than in ours: `cu_seqlens_*_start *
> stride_*` is a cumulative token offset across the whole batch, which crosses `2**31`
> even for contiguous tensors (~300k tokens at 56 heads × 128 dim).
>
> ### Fix
>
> Promote the index to int64 where it is defined, exactly as `off_z` / `off_h` already
> are. 28 casts in 7 files; PR to follow.
>
> ### Out of scope, but adjacent
>
> `csrc/qattn/qk_int_sv_f8_cuda_sm89.cuh` and `csrc/fused/fused.cu` compute offsets in
> `uint32_t`, which gives a `2**32` ceiling instead of `2**31`. Those kernels only see
> freshly allocated contiguous tensors, so we did not hit it, but the same class of
> limit is there.
